spring-integration-core/src/main/java/org/springframework/integration/transaction/DefaultTransactionSynchronizationFactory.java

```diff
@@ -46,13 +49,16 @@ public DefaultTransactionSynchronizationFactory(TransactionSynchronizationProces
 	@Override
 	public TransactionSynchronization create(Object key) {
 		Assert.notNull(key, "'key' must not be null");
-		DefaultTransactionalResourceSynchronization synchronization = new DefaultTransactionalResourceSynchronization(key);
-		TransactionSynchronizationManager.bindResource(key, synchronization.getResourceHolder());
-		return synchronization;
+		if (!TransactionSynchronizationManager.hasResource(key)) {
```

Comment:
I think it would be cleaner to move this test to the callers rather than this strange return `null` if something's already bound with the key.